### PR TITLE
cephadm: Make sysctl conf writing idempotent

### PR DESCRIPTION
**Pull request by zmc**
By checking the system's actual settings against our desired settings
before writing out the config file.

Fixes: https://tracker.ceph.com/issues/56026

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by zmc** [comment]
Filed #46649 so that we can have quicker feedback on changes like these

File: src/cephadm/cephadm [hunk 1]
```diff
@@ -3548,6 +3549,33 @@ def install_sysctl(ctx: CephadmContext, fsid: str, daemon_type: str) -> None:
...
+    out, err, code = call(ctx, ['sysctl', '-b', variable])
+    return out or None
+
+
+def filter_sysctl_settings(ctx: CephadmContext, lines: List[str]) -> List[str]:
+    """
+    Given a list of sysctl settings, examine the system's current configuration
+    and return those which are not currently set as described.
+    """
+    def testSetting(desiredLine: str) -> bool:
```
Comment (rkachach) [inline src/cephadm/cephadm:3565]: `lower_case_with_underscores` is preffered both for function names and variables as recommended by PEP8 and to keep consistency with the rest of the file style in this case.

File: src/cephadm/cephadm [hunk 2]
```diff
@@ -3548,6 +3549,33 @@ def install_sysctl(ctx: CephadmContext, fsid: str, daemon_type: str) -> None:
...
+        commentStart = desiredLine.find('#')
+        if commentStart != -1:
+            desiredLine = desiredLine[:commentStart]
+        desiredLine = desiredLine.strip()
+        if not desiredLine:
+            return False
+        setting, desiredValue = map(lambda s: s.strip(), desiredLine.split('='))
+        actualValue = sysctl_get(ctx, setting)
+        return desiredValue != actualValue
+    return list(filter(testSetting, lines))
```
Comment (rkachach) [inline src/cephadm/cephadm:3576]: consider using python3 style: `[testSetting(line) for line in lines]`

File: src/cephadm/tests/test_cephadm.py [hunk 3]
```diff
@@ -2323,4 +2323,34 @@ def test_ip_in_subnet(self):
...
-            rc = cd.ip_in_sublets('fe80:2030:31:24', 'fe80::/64')
\ No newline at end of file
+            rc = cd.ip_in_sublets('fe80:2030:31:24', 'fe80::/64')
+
+
+class TestSysctl:
+    @mock.patch('cephadm.sysctl_get')
+    def test_filter_sysctl_settings(self, sysctl_get):
+        ctx = cd.CephadmContext()
+        input = [
```
Comment (rkachach) [inline src/cephadm/tests/test_cephadm.py:2333]: It would be good to test against lines with spaces chars (spaces and/or tabs).

**Review by rkachach — CHANGES_REQUESTED** [review]
Thanks for the contribution, LGTM, just some minor suggestions

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2022-06-14_00:02:10-orch:cephadm-wip-adk-testing-2022-06-13-1707-distro-default-smithi/

1 failure tracked by https://tracker.ceph.com/issues/55986

File: src/cephadm/cephadm (hunk 2 above)
Comment (zmc) [reply src/cephadm/cephadm]: that would yield a list of booleans

File: src/cephadm/cephadm [hunk 4]
```diff
@@ -3548,6 +3549,33 @@ def install_sysctl(ctx: CephadmContext, fsid: str, daemon_type: str) -> None:
         call_throws(ctx, ['sysctl', '--system'])
 
 
+def sysctl_get(ctx: CephadmContext, variable: str) -> Union[str, None]:
+    """
+    Read a sysctl setting by executing 'sysctl -b {variable}'
+    """
+    out, err, code = call(ctx, ['sysctl', '-b', variable])
```
[the remark this answers is not in the document]
Comment (zmc) [reply src/cephadm/cephadm:3556]: `call()` does this as-is; do you think we need logging with a different appearance or something?

File: src/cephadm/cephadm (hunk 1 above)
Comment (zmc) [reply src/cephadm/cephadm]: you're absolutely right - been writing too much JS lately apparently

**Review by zmc — COMMENTED** [review]
Repushed with some changes: renamed variables, added some whitespace-related checks, and added some unit tests.

File: src/cephadm/cephadm (hunk 4 above)
[the remark this answers is not in the document]
Comment (rkachach) [reply src/cephadm/cephadm:3556]: you are right :) ... my bad, I though you was calling the python native call.

File: src/cephadm/cephadm [hunk 5]
```diff
@@ -3548,6 +3549,35 @@ def install_sysctl(ctx: CephadmContext, fsid: str, daemon_type: str) -> None:
...
+
+
+def filter_sysctl_settings(ctx: CephadmContext, lines: List[str]) -> List[str]:
+    """
+    Given a list of sysctl settings, examine the system's current configuration
+    and return those which are not currently set as described.
+    """
+    def test_setting(desired_line: str) -> bool:
+        # Remove any comments
+        commentStart = desired_line.find('#')
```
Comment (rkachach) [inline src/cephadm/cephadm:3567]: it seems that `commentStart` escaped the vars renaming :)

File: src/cephadm/cephadm (hunk 2 above)
Comment (rkachach) [reply src/cephadm/cephadm]: yeah, you are right, actually it should be: [l for l in lines if testSetting(l)]

File: src/cephadm/tests/test_cephadm.py [hunk 6]
```diff
@@ -2323,4 +2323,40 @@ def test_ip_in_subnet(self):
...
+            "# just a comment",
+            # As should whitespace-only lines",
+            "   \t ",
+            "   =  \t  ",
+            # inline comments are stripped when querying
+            "something = value # inline comment",
+            "fs.aio-max-nr = 1048576",
+            "kernel.pid_max = 4194304",
+            "vm.lowmem_reserve_ratio = 256\t256\t32\t0\t0",
+            "  vm.max_map_count       =            65530    ",
```
Comment (rkachach) [inline src/cephadm/tests/test_cephadm.py:2344]: Just a simple suggestion, plz ignore if you think the coverage is already good :)

plz consider including this line `vm.max_map_count       =            xxxxx ` in the desired output (by changing its value), this way we make sure the code is really filtering and not only returning always the first value. So the output should be:

```
assert result == [
            "fs.aio-max-nr = 1048576",
             "vm.max_map_count = xxxxx"
        ]

```

Comment (zmc) [reply src/cephadm/tests/test_cephadm.py:2344]: Added one more just for you 😄

**Review by rkachach — APPROVED** [review]

**Comment by zmc** [comment]
When PRs are blocked due to flaky Jenkins jobs, what's the normal procedure? We seem to have quite a lot of those right now.

**Comment by adk3798** [comment]
> When PRs are blocked due to flaky Jenkins jobs, what's the normal procedure? We seem to have quite a lot of those right now.

we just have to rerun until they pass unfortunately. Usually, if a job is failing often, someone is looking at why and it gets resolved, but when you're trying to get something tested when a job is still in a flaky state it's usually reruns. Can't merge otherwise.

**Comment by adk3798** [comment]
jenkins test api

**Comment by zmc** [comment]
jenkins test windows

**Comment by zmc** [comment]
jenkins test dashboard cephadm

**Pull request merged**